### PR TITLE
[ST-Engine] Add Submit Job From Client To Server

### PR DESCRIPTION
**Pull request by EricJoy2048**
subtask of https://github.com/apache/incubator-seatunnel/issues/2260
fix https://github.com/apache/incubator-seatunnel/issues/2333

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java [hunk 1]
```diff
@@ -63,22 +67,26 @@ public class JobConfigParser {
...
     }
 
-    public List<Action> parse() {
+    public ImmutablePair<List<Action>, Set<URL>> parse() {
         Config seaTunnelJobConfig = new ConfigBuilder(Paths.get(jobDefineFilePath)).getConfig();
         List<? extends Config> sinkConfigs = seaTunnelJobConfig.getConfigList("sink");
         List<? extends Config> transformConfigs = seaTunnelJobConfig.getConfigList("transform");
         List<? extends Config> sourceConfigs = seaTunnelJobConfig.getConfigList("source");
 
         if (sinkConfigs.size() == 1 && sourceConfigs.size() == 1 & transformConfigs.size() <= 1) {
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java]: Is there a risk of NPE?

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/ConnectorInstanceLoader.java [hunk 2]
```diff
@@ -31,16 +31,24 @@
 
 import org.apache.seatunnel.shade.com.typesafe.config.Config;
 
+import com.google.common.collect.Lists;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import java.net.URL;
+import java.util.List;
+
 import scala.Serializable;
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/ConnectorInstanceLoader.java:41]: Utility class should not be instantiated, it's better to add
```
private ConnectorInstanceLoader() {
    throw new IllegalStateException("Utility class");
  }
```

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java [hunk 3]
```diff
@@ -211,43 +230,57 @@ private void initRelationMap(List<? extends Config> sourceConfigs, List<? extend
...
         SourceAction sourceAction =
-            new SourceAction(idGenerator.getNextId(), seaTunnelSource.getPluginName(), seaTunnelSource);
+            createSourceAction(idGenerator.getNextId(), pair.getLeft().getPluginName(), pair.getLeft(),
+                pair.getRight());
         sourceAction.setParallelism(getSourceParallelism(sourceConfigs.get(0)));
+
+        ImmutablePair<SeaTunnelSink<SeaTunnelRow, Serializable, Serializable, Serializable>, List<URL>>
+            sinkListImmutablePair;
+
         if (transformConfigs.size() != 0) {
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java]: using Collection.isEmpty() makes the code more readable and can be more performant. The time complexity of any isEmpty() method implementation should be O(1) whereas some implementations of size() can be O(n).

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java [hunk 4]
```diff
@@ -211,43 +230,57 @@ private void initRelationMap(List<? extends Config> sourceConfigs, List<? extend
...
+                sinkListImmutablePair.getLeft().getPluginName(),
                 Lists.newArrayList(transformAction),
-                seaTunnelSink
+                sinkListImmutablePair.getLeft(),
+                sinkListImmutablePair.getRight()
             );
             sinkAction.setParallelism(transformAction.getParallelism());
-            return Lists.newArrayList(sinkAction);
+            actions.add(sinkAction);
         } else {
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/JobConfigParser.java]: We can return directly, so no `else` judgment is needed.

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public CompletableFuture<Void> requestOnMasterAndGetCompletableFuture(@NonNull ClientMessage request) {
+        UUID masterUuid = hazelcastClient.getClientClusterService().getMasterMember().getUuid();
+        return requestAndGetCompletableFuture(masterUuid, request);
+    }
+
+    public CompletableFuture<Void> requestOnAnyMemberAndGetCompletableFuture(@NonNull ClientMessage request) {
+        UUID masterUuid = hazelcastClient.getClientClusterService().getMasterMember().getUuid();
+        return requestAndGetCompletableFuture(masterUuid, request);
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java]: Seems like they do the same thing,

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    public <S> S requestAndDecodeResponse(UUID uuid, ClientMessage request,
+                                          Function<ClientMessage, Object> decoder) {
+        ClientInvocation invocation = new ClientInvocation(hazelcastClient, request, null, uuid);
+        try {
+            ClientMessage response = invocation.invoke().get();
+            return serializationService.toObject(decoder.apply(response));
+        } catch (Throwable t) {
+            throw ExceptionUtil.rethrow(t);
+        }
```
Comment (CalvinKirs) [inline seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java]: invocation.invoke().get() may throw an InterruptedException.
we don't ignore this.you need restore interrupted state

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/ConnectorInstanceLoader.java (hunk 2 above)
Comment (EricJoy2048) [reply seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/ConnectorInstanceLoader.java:41]: done

File: seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java (hunk 6 above)
Comment (EricJoy2048) [reply seatunnel-engine/seatunnel-engine-client/src/main/java/org/apache/seatunnel/engine/client/SeaTunnelHazelcastClient.java]: done

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**